src/v/kafka/server/handlers/delete_records.cc

```diff
@@ -0,0 +1,248 @@
+ * by the Apache License, Version 2.0
+ */
+
+#include "kafka/server/handlers/delete_records.h"
+
+#include "cluster/metadata_cache.h"
+#include "cluster/partition_manager.h"
+#include "cluster/shard_table.h"
+#include "model/fundamental.h"
+
+#include <vector>
+
+namespace kafka {
+
+/// Returned in responses where kafka::error_code is anything else then a value
+/// of error_code::none
+static constexpr auto invalid_low_watermark = model::offset(-1);
+
+/// Compare against user provided value of truncation offset, this value will
+/// indicate to truncate at  the current partition high watermark
+static constexpr auto current_high_watermark = model::offset(-1);
+
+static std::vector<delete_records_partition_result>
+make_partition_errors(const delete_records_topic& t, error_code ec) {
+    std::vector<delete_records_partition_result> r;
+    for (const auto& p : t.partitions) {
+        r.push_back(delete_records_partition_result{
+          .partition_index = p.partition_index,
+          .low_watermark = invalid_low_watermark,
+          .error_code = ec});
+    }
+    return r;
+}
+
+/// Performs validation of topics, any failures will result in a list of
+/// partitions that all contain the identical error codes
+static std::vector<delete_records_partition_result>
+validate_at_topic_level(request_context& ctx, const delete_records_topic& t) {
+    const auto is_authorized = [&ctx](const delete_records_topic& t) {
+        return ctx.authorized(security::acl_operation::remove, t.name);
+    };
+    const auto is_deletable = [&ctx](const delete_records_topic& t) {
+        const auto cfg = ctx.metadata_cache().get_topic_cfg(
+          model::topic_namespace_view(model::kafka_namespace, t.name));
+        if (!cfg || !cfg->properties.cleanup_policy_bitflags) {
+            return false;
+        }
+        /// TODO: call is_collectable()? Maybe theres overrides
+        const auto flags = cfg->properties.cleanup_policy_bitflags;
+        return (*flags & model::cleanup_policy_bitflags::deletion)
+               == model::cleanup_policy_bitflags::deletion;
+    };
+    const auto is_cloud_enabled = [&ctx](const delete_records_topic& t) {
+        const auto cfg = ctx.metadata_cache().get_topic_cfg(
+          model::topic_namespace_view(model::kafka_namespace, t.name));
+        if (!cfg) {
+            return false;
+        }
+        const auto si_flags = cfg->properties.shadow_indexing;
+        return si_flags && *si_flags != model::shadow_indexing_mode::disabled;
+    };
+    const auto is_internal_topic = [](const delete_records_topic& t) {
+        const auto& internal_topics
+          = config::shard_local_cfg().kafka_nodelete_topics();
+        return std::find_if(
+                 internal_topics.begin(),
+                 internal_topics.end(),
+                 [t](const ss::sstring& name) { return name == t.name; })
+               != internal_topics.end();
+    };
```

Comment:
Here naming is a bit inconsistent, i.e. method is named `is_internal_topic` but the check validates if a topic is on non delete list